src/tools/rbd_mirror/image_map/Policy.cc

```diff
@@ -0,0 +1,239 @@
+}
+
+template <typename I>
+std::string Policy<I>::_lookup(const std::string &global_image_id) {
+  assert(m_map_lock.is_locked());
+
+  for (auto it = m_map.begin(); it != m_map.end(); ++it) {
+    if (it->second.find(global_image_id) != it->second.end()) {
+      return it->first;
+    }
+  }
+
+  return UNMAPPED_INSTANCE_ID;
+}
+
+template <typename I>
+std::string Policy<I>::lookup(const std::string &global_image_id) {
+  dout(20) << ": global_image_id=" << global_image_id << dendl;
+
+  RWLock::RLocker map_lock(m_map_lock);
+  return _lookup(global_image_id);
+}
+
+template <typename I>
+std::string Policy<I>::map(const std::string &global_image_id) {
+  dout(20) << ": global_image_id=" << global_image_id << dendl;
+
+  RWLock::WLocker map_lock(m_map_lock);
+
+  std::string instance_id = do_map(global_image_id);
+  _map(global_image_id, instance_id);
+
+  return instance_id;
+}
+
+template <typename I>
+bool Policy<I>::unmap(const std::string &global_image_id) {
+  dout(20) << ": global_image_id=" << global_image_id << dendl;
+
+  RWLock::WLocker map_lock(m_map_lock);
+
+  std::string instance_id = _lookup(global_image_id);
+  if (instance_id == UNMAPPED_INSTANCE_ID) {
+    return false;
+  }
+
+  m_map[instance_id].erase(global_image_id);
+  m_shuffled_timestamp.erase(global_image_id);
+  return true;
+}
+
+template <typename I>
+void Policy<I>::remap(const std::string &global_image_id,
+                      const std::string &source_instance_id,
+                      const std::string &target_instance_id, utime_t remap_time) {
+  dout(20) << ": global_image_id=" << global_image_id << ", source_instance_id="
+           << source_instance_id << ", target_instance_id=" << target_instance_id
+           << dendl;
+
+  assert(m_map_lock.is_wlocked());
+
+  m_map[target_instance_id].emplace(global_image_id);
+  m_map[source_instance_id].erase(global_image_id);
+  m_shuffled_timestamp[global_image_id] = remap_time;
+}
+
+template <typename I>
+bool Policy<I>::instance_present(const std::string &instance_id) {
+  dout(20) << ": instance_id=" << instance_id << dendl;
+
+  RWLock::RLocker map_lock(m_map_lock);
+  return m_map.find(instance_id) != m_map.end();
+}
+
+template <typename I>
+void Policy<I>::add_instances(const std::vector<std::string> &instances) {
+  dout(20) << dendl;
+
+  assert(m_map_lock.is_wlocked());
+
+  for (auto const &instance : instances) {
+    dout(10) << ": adding instance_id=" << instance << dendl;
+    m_map.emplace(instance, std::set<std::string>{});
+  }
+}
+
+template <typename I>
+void Policy<I>::remove_instances(const std::vector<std::string> &instances) {
+  dout(20) << dendl;
+
+  assert(m_map_lock.is_wlocked());
+
+  for (auto const &instance : instances) {
+    dout(10) << ": removing instance_id=" << instance << dendl;
+    if (!m_map[instance].empty()) {
+      dout(5) << ": instance_id=" << instance << " still has " << m_map[instance].size()
```

Comment:
Would this be a bug?

Reply:
It would be if the policy implementation does not (for some reason) remap all images from the instances that are being removed. I did not treat it as a hard error here. With the changes in `update_images_removed()` (ignoring if an image is not found in the map), we should probably `assert()` here.

Reply:
Sounds good to me